lib/THC/THCTensorMath.cuh

```diff
@@ -23,4 +23,80 @@ __global__ void THCTensor_copyToDiagonal(T* a, T* b, ptrdiff_t start, ptrdiff_t
   }
 }
 
+#define CAT_ARRAY_BATCH_SIZE 1024
+#define CAT_ARRAY_MAX_INPUT_DIMS 4
+
+// Similar to any other IndexToOffset calculation for copying along a given dimension.
+template <typename IndexType, int Dims>
+struct CatArrIndexToOffset {
+  static inline __device__ IndexType compute(
+      const IndexType outputSize[Dims],
+      const IndexType outputStride[Dims],
+      const IndexType dimSize,
+      const unsigned int concatDim,
+      IndexType linearIndex) {
+    IndexType offset = 0;
+
+#pragma unroll
+    for (int i = Dims - 1; i >= 0; --i) {
+      IndexType curDimSize = i == concatDim ? dimSize : outputSize[i];
+      IndexType curDimIndex = linearIndex % curDimSize;
+      IndexType curDimOffset = curDimIndex * outputStride[i];
+      offset += curDimOffset;
+
+      linearIndex /= curDimSize;
```

Comment:
I am also not sure if there is any point for the loop to go down to 0. Unless I am missing something, It should theoretically be sufficient to go down to 1 and return `offset + linearIndex * outputStride[0]` which would behave like a special case when `Dims == 1`